pkg/cmd/pulumi/import.go

```diff
@@ -0,0 +1,252 @@
+				Config:    cfg.Config,
+				Decrypter: cfg.Decrypter,
+				Snapshot:  snap,
+			}
+			cwd, err := os.Getwd()
+			if err != nil {
+				return err
+			}
+			sink := diag.DefaultSink(os.Stdout, os.Stderr, diag.FormatOptions{Pwd: cwd, Color: colors.Always})
+			ctx, err := plugin.NewContext(sink, sink, nil, target, cwd, nil, nil)
+			if err != nil {
+				return err
+			}
+			host := ctx.Host
+
+			// Load the default provider.
+			//
+			// TODO: allow a provider to be specified on the command line or pull from the snapshot, as we can't
+			// otherwise store to the state.
+			pkg := urn.Type().Package()
+			prov, err := host.Provider(pkg, nil)
+			if err != nil {
+				return err
+			}
+			pkgCfg, err := target.GetPackageConfig(pkg)
+			if err != nil {
+				return err
+			}
+			inputs := resource.PropertyMap{}
+			for k, v := range pkgCfg {
+				inputs[resource.PropertyKey(k.Name())] = resource.NewStringProperty(v)
+			}
+			if err = prov.Configure(inputs); err != nil {
+				return err
+			}
+
+			// Grab the resource's ID.
+			id := resource.ID(args[2])
+
+			var initErrors []string
+			read, _, err := prov.Read(urn, id, nil, nil)
+			if err != nil {
+				if initErr, isInitErr := err.(*plugin.InitError); isInitErr {
+					initErrors = initErr.Reasons
+				} else {
+					return err
+				}
+			}
+			if read.Outputs == nil {
+				return errors.Errorf("resource '%v' does not exist", id)
+			}
+			if read.Inputs == nil {
+				return errors.Errorf(
+					"provider does not support importing resources; please try updating the '%v' plugin",
+					urn.Type().Package())
+			}
+
+			stackURN := resource.URN("")
+			if stackResource != nil {
+				stackURN = stackResource.URN
+			}
+
+			state := resource.NewState(urn.Type(), urn, true, false, id, resource.PropertyMap{}, read.Outputs,
+				stackURN, true, false, nil, initErrors, "", nil, false, nil, nil, nil, "")
+
+			// Magic up an old state so we can perform a diff.
+			old := resource.NewState(state.Type, state.URN, state.Custom, false, state.ID, read.Inputs, read.Outputs,
+				state.Parent, state.Protect, false, state.Dependencies, state.InitErrors, state.Provider,
+				state.PropertyDependencies, false, nil, nil, &state.CustomTimeouts, state.ImportID)
+
+			// Diff the user inputs against the provider inputs. If there are any differences, fail the import.
+			diff, err := deploy.DiffResource(state.URN, state.ID, old.Inputs, old.Outputs, state.Inputs, prov, false, nil)
+			if err != nil {
+				return err
+			}
+
+			if diff.Changes != plugin.DiffNone {
+				for path, pdiff := range diff.DetailedDiff {
+					elements, err := resource.ParsePropertyPath(path)
+					if err != nil {
+						continue
+					}
+
+					source := old.Outputs
+					if pdiff.InputDiff {
+						source = old.Inputs
+					}
+
+					if old, ok := elements.Get(resource.NewObjectProperty(source)); ok {
+						// Ignore failure here.
+						elements.Add(resource.NewObjectProperty(state.Inputs), old)
+					}
+				}
+			}
```

Comment:
This is the biggest open question I have.

This code matches the steps a user would need to take to author a resource definition that will successfully import, but it ends up leaving out anything that might have been computed. I _think_ that this is correct (and is probably the only way for this to work successfully, as outputs can be set for properties that would conflict were they inputs), but it can result in surprisingly small resource definitions.

Reply:
>  it ends up leaving out anything that might have been computed

I'm not sure if this is the same point - but it seems to also leave out anything that is "optional". 

I just tried for example:

```
➜  tsimporttest PULUMI_EXPERIMENTAL=true /opt/pulumi/bin/pulumi import aws:cloudformation/stack:Stack s example-oidc-iam-sa-13d5faa9               


import * as pulumi from "@pulumi/pulumi";
import * as aws from "@pulumi/aws";

const s = new aws.cloudformation.Stack("s", {
    name: "example-oidc-iam-sa-13d5faa9",
    tags: {
        Name: "example-oidc-iam-sa-nodes",
    },
});
```

That is enough to import the resource, but is definitely not enough to re-create the resource.  Indeed - trying to create a resources from that fails with:

```
  aws:cloudformation:Stack (s):
    error: Creating CloudFormation stack failed: ValidationError: Either Template URL or Template Body must be specified.
        status code: 400, request id: 503c81d7-ee0e-45d4-adee-edbcf63b9482
```

The `outputs` section in the Pulumi stack does contain the desired `templateBody` property, but it wasn't brought over to the inputs.

Reply:
The problem here is that `template_body` is optional or computed. It does come back in the inputs, but because it's a computed property, the provider sees no diff if the property is not present in the inputs.

I should point out that the generated definitions are all incomplete because they do not include the `import` resource option.

```go
			"template_body": {
				Type:         schema.TypeString,
				Optional:     true,
				Computed:     true,
				ValidateFunc: validateCloudFormationTemplate,
				StateFunc: func(v interface{}) string {
					template, _ := normalizeCloudFormationTemplate(v)
					return template
				},
			},
			"template_url": {
				Type:     schema.TypeString,
				Optional: true,
			},
```

Reply:
It might be useful to see what the experience is if we don’t leave off Computed. (As long as they are Optional or Required). 

The current experience of underspecifying feels less “correct” than overspecifying.  Though it would be interesting to hand validate a number of concrete cases to better characterize the trade off.

Reply:
> It might be useful to see what the experience is if we don’t leave off Computed. (As long as they are Optional or Required).

Anything that is optional or required does come back in the inputs; we just remove them when we process the result of the diff. End-to-end the process looks like this:

1. Read the current inputs and state from the provider
2. Diff an empty set of inputs against the current inputs + state
3. If there are diffs, copy the differing properties into the inputs
4. Generate the resource's definition using the inputs from (3)

When I originally wrote this up, I did not do (2) and (3). Instead, I used the inputs from (1) in (4). This resulted in very large definitions that at least in some cases--probably `aws:ec2:Instance` or `aws:s3:Bucket`--did not successfully `Check` due to the presence of conflicting input properties. For example, this can happen if there are two in/out properties A and B where if A is an input B is computed and vice-versa. The experience of copy-pasting a definition that didn't successfully import (and would not successfully create) is what led me to the current approach.

> The current experience of underspecifying feels less “correct” than overspecifying. Though it would be interesting to hand validate a number of concrete cases to better characterize the trade off.

FWIW, the current experience is consistent with what a user would need to do manually in order for an import to succeed.